llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -341,8 +338,23 @@ void AMDGPUDAGToDAGISel::PreprocessISelDAG() {
 
     switch (N->getOpcode()) {
     case ISD::BUILD_VECTOR:
-      // TODO: Match load d16 from shl (extload:i16), 16
-      MadeChange |= matchLoadD16FromBuildVector(N);
+      // D16 optimization requires subtarget support
```

Comment:
This is a weird place to put the optimization. This doesn't have the special generalized RMW property that the d16 loads do.

 Buffers don't just come out of nowhere, you could do this on the IR.

Reply:
Isn't that for the legacy intrinsics? The new intrinsics have the explicit soffset and voffset operands

Reply:
No

Even with the new ones, if I have `/*voffset=*/%x + 16` in the IR, ISel will pattern-match that `+ 16` part into `offset:16` on the instruction

Reply:
So I claim that voffset => soffset should either go there or be an IR pass after uniformity analysis like you suggested.

(Making it an IR pass also means you can do s_buffer_load promotion there)